core/src/block.rs

```diff
@@ -19,894 +19,614 @@ use iroha_data_model::{block::*, events::prelude::*, transaction::prelude::*};
+}
 
-        /// Chain block with the existing blockchain.
-        pub fn chain(
-            self,
-            height: u64,
-            previous_block_hash: Option<HashOf<VersionedCommittedBlock>>,
-            view_change_index: u64,
-            committed_with_topology: Topology,
-        ) -> ChainedBlock {
-            ChainedBlock {
-                transactions: self.transactions,
-                event_recommendations: self.event_recommendations,
-                header: BlockHeader {
-                    timestamp: self.timestamp,
-                    consensus_estimation: DEFAULT_CONSENSUS_ESTIMATION_MS,
-                    height: height + 1,
-                    view_change_index,
-                    previous_block_hash,
-                    transactions_hash: None,
-                    rejected_transactions_hash: None,
-                    committed_with_topology: committed_with_topology.sorted_peers,
-                },
-            }
-        }
+impl PendingBlock {
+    /// Create a new [`PendingBlock`] from transactions.
+    #[inline]
+    pub fn new(
+        transactions: Vec<VersionedAcceptedTransaction>,
+        event_recommendations: Vec<Event>,
+        height: u64,
+        previous_block_hash: Option<HashOf<VersionedCommittedBlock>>,
+        view_change_index: u64,
+        committed_with_topology: Topology,
+        key_pair: KeyPair,
+        transaction_validator: &TransactionValidator,
+        wsv: &WorldStateView,
+    ) -> Self {
+        #[allow(clippy::expect_used)]
+        let timestamp = crate::current_time().as_millis();
+        // TODO: Need to check if the `transactions` vector is empty. It shouldn't be allowed.
+
+        let mut header = BlockHeader {
+            timestamp,
+            consensus_estimation: DEFAULT_CONSENSUS_ESTIMATION_MS,
+            height,
+            view_change_index,
+            previous_block_hash,
+            transactions_hash: None,
+            rejected_transactions_hash: None,
+            committed_with_topology: committed_with_topology.sorted_peers,
+        };
 
-        /// Create a new blockchain with current block as a first block.
-        pub fn chain_first_with_topology(self, genesis_topology: Topology) -> ChainedBlock {
-            ChainedBlock {
-                transactions: self.transactions,
-                event_recommendations: self.event_recommendations,
-                header: BlockHeader {
-                    timestamp: self.timestamp,
-                    consensus_estimation: DEFAULT_CONSENSUS_ESTIMATION_MS,
-                    height: 1,
-                    view_change_index: 0,
-                    previous_block_hash: None,
-                    transactions_hash: None,
-                    rejected_transactions_hash: None,
-                    committed_with_topology: genesis_topology.sorted_peers,
-                },
+        let mut txs = Vec::new();
+        let mut rejected = Vec::new();
+
+        let wsv = wsv.clone();
```

Comment:
Consider making this function accept a `WorldStateView` and cloning one level above.